File: C++/myLeetCode/Array/209. Minimum Size Subarray Sum.cpp
// 209. Minimum Size Subarray Sum
// https://leetcode.com/problems/minimum-size-subarray-sum/

#include <iostream>
#include <limits>
#include <vector>

class Solution {
  public:
    int minSubArrayLen(int target, std::vector<int> &nums) {
        int result = std::numeric_limits<int>::max();
        int sum = 0;
        int left = 0;
        int subLength = 0;
        int n = nums.size();
        for (int right = 0; right < n; right++) {
            sum += nums[right];
            while (sum >= target) {
                subLength = (right - left + 1);
                result = std::min(result, subLength);
                sum -= nums[left++];
            }
        }
        return result == std::numeric_limits<int>::max() ? 0 : result;
    }
};

int main() {
    Solution sol;
    std::vector<int> nums = {2, 3, 1, 2, 4, 3};
    int target = 7;
    int ans = sol.minSubArrayLen(target, nums);
    std::cout << ans << std::endl;
    return 0;
}

File: C#/myLeetCode/Array/209. Minimum Size Subarray Sum.cs
// 209. Minimum Size Subarray Sum
// https://leetcode.com/problems/minimum-size-subarray-sum/

using System;

public class Solution {
    public int MinSubArrayLen(int target, int[] nums) {
        int minLength = int.MaxValue;
        int left = 0;
        int currentSum = 0;

        for (int right = 0; right < nums.Length; right++) {
            currentSum += nums[right];

            while (currentSum >= target) {
                minLength = Math.Min(minLength, right - left + 1);
                currentSum -= nums[left];
                left++;
            }
        }

        return minLength == int.MaxValue ? 0 : minLength;
    }
}

public class Program {
    public static void Main() {
        Solution solution = new Solution();
        
        int[] nums = {2,3,1,2,4,3};
        int target = 7;

        int result = solution.MinSubArrayLen(target, nums);
        Console.WriteLine("The minimum length of the subarray with a sum greater than or equal to " + target + " is: " + result);
    }
}
